utils/estimated_histogram.hh

```diff
@@ -47,13 +47,359 @@
 #include <chrono>
 #include <seastar/core/metrics_types.hh>
 #include <seastar/core/print.hh>
-
-#include <seastar/core/print.hh>
-
 #include "seastarx.hh"
+#include <seastar/core/bitops.hh>
+#include <limits>
 
 namespace utils {
 
+// Returns largest N such that 2^N <= v.
+// Undefined for v == 0.
+inline constexpr size_t pow2_rank(uint64_t v) {
+    return std::numeric_limits<uint64_t>::digits - 1 - count_leading_zeros(v);
+}
+
+template<uint64_t Min, uint64_t Max, size_t NumBuckets>
+class pseudo_floatingpoint_histogram {
+    std::array<uint64_t, NumBuckets> buckets;
+public:
+    /**
+     * This is a pseudo floating point implementation of an estimated histogram.
+     * When entering a value:
+     * All values lower than the MIN will be included in the first bucket.
+     * All values higher than MAX will be included the last bucket that serves as the
+     * infinity bucket.
+     *
+     * buckets are distributed as pseudo floating point:
+     * The range [MIN, MAX) is split into log2 ranges.
```

Comment:
This documentation should be moved to the class header, not be here in the middle of the class.

Reply:
ok